### PR TITLE
NO OP and QN Reference Documentation

### PR DESCRIPTION
**Pull request by bug-sam**
#6 
#7

### CODE REVIEW

File: no_op.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    - creates ProcessorJob
+    - creates an OriginalFile and points to a file with a header
+    - creates a test Sample
+    - assaciates the ProcessorJob and OriginalFile to the test Sample
+- runs the no op pipeline using the above prerequisites
+- creates prerequisites for the second pipeline run
+    - creates ProcessorJob
+    - creates an OriginalFile and points to a file with no header
+    - creates a test Sample
+    - assaciates the ProcessorJob and OriginalFile to the test Sample
```
Comment (kurtwheeler) [inline no_op.md]:
```suggestion
    - associates the ProcessorJob and OriginalFile to the test Sample
```

File: no_op.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+### `test_convert_processed_illumina`
+
+Tests the entire no op pipeline for an Illumina gene sample
+
+test steps:
+- creates prerequisites
+    - creates ProcessorJob
+    - creates an OriginalFile and points to a file with a header
+    - creates a test Sample
+    - assaciates the ProcessorJob and OriginalFile to the test Sample
```
Comment (kurtwheeler) [inline no_op.md]:
```suggestion
    - associates the ProcessorJob and OriginalFile to the test Sample
```

File: no_op.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+### `test_convert_illumina_no_header`
+
+Tests the entire no op pipeline for an Illumina gene sample without a header
+
+test steps:
+- creates prerequisites
+    - creates ProcessorJob
+    - creates an OriginalFile and points to a file without a header
+    - creates a test Sample
+    - assaciates the ProcessorJob and OriginalFile to the test Sample
```
Comment (kurtwheeler) [inline no_op.md]:
```suggestion
    - associates the ProcessorJob and OriginalFile to the test Sample
```

File: no_op.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+### `test_convert_illumina_bad_cols`
+
+Tests the entire no op pipeline for an Illumina gene sample with a bad column in the input file
+
+test steps:
+- creates prerequisites
+    - creates ProcessorJob
+    - creates an OriginalFile and points to a file without a header
+    - creates a test Sample
+    - assaciates the ProcessorJob and OriginalFile to the test Sample
```
Comment (kurtwheeler) [inline no_op.md]:
```suggestion
    - associates the ProcessorJob and OriginalFile to the test Sample
```

File: no_op.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+- runs the no op pipeline using the above prerequisites
+- asserts the job successfully ran
+- asserts that the output file exists
+- asserts that the output file is the size we expect
+    - within 5% since the output file sizes can vary each run
+
+### `test_convert_illumina_bad_cols`
+
+Tests the entire no op pipeline for an Illumina gene sample with a bad column in the input file
```
Comment (kurtwheeler) [inline no_op.md]:
```suggestion
Tests the entire No Op pipeline for an Illumina gene sample with a bad column in the input file
```

File: no_op.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+- asserts that the output file exists
+- asserts that the output file is the size we expect
+    - within 5% since the output file sizes can vary each run
+- asserts that the output file passes quality control
+    - makes sure each row is no more than 2 columns
+    - if there's more than 2 column in a row that's considered bad data likely caused by an old machine or bad processing
+
+### `test_convert_illumina_no_header`
+
+Tests the entire no op pipeline for an Illumina gene sample without a header
```
Comment (kurtwheeler) [inline no_op.md]:
```suggestion
Tests the entire No Op pipeline for an Illumina gene sample without a header
```

File: no_op.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    - assaciates the ProcessorJob and OriginalFile to the test Sample
+- runs the no op pipeline using the above prerequisites
+- asserts the job successfully ran
+- asserts that the output file exists
+- asserts that the output file is the size we expect
+    - within 5% since the output file sizes can vary each run
+
+### `test_convert_processed_illumina`
+
+Tests the entire no op pipeline for an Illumina gene sample
```
Comment (kurtwheeler) [inline no_op.md]:
```suggestion
Tests the entire No Op pipeline for an Illumina gene sample
```

File: no_op.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
+# No Op
+
+## Tests
+
+
+### `test_convert_simple_pcl`
+
+Tests the entire no op pipeline for an AFFY gene sample
```
Comment (kurtwheeler) [inline no_op.md]:
```suggestion
Tests the entire No Op pipeline for an Affy gene sample
```

File: no_op.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+Tests the entire no op pipeline for an Illumina gene sample with a bad column in the input file
+
+test steps:
+- creates prerequisites
+    - creates ProcessorJob
+    - creates an OriginalFile and points to a file without a header
+    - creates a test Sample
+    - assaciates the ProcessorJob and OriginalFile to the test Sample
+- runs the no op pipeline using the above prerequisites
```
Comment (kurtwheeler) [inline no_op.md]: There's a few more places `no op` is uncapitalized. I think standardizing on capitalized makes the most sense.

File: qn_reference.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+Tests the entire QN Reference pipeline
+
+Also creates a smasher job that uses the output from the QN Reference run and tests that
+
+test steps:
+- creates prerequisites
+    - creates ProcessorJob
+    - creates test Organism
+    - creates test experiment
+    - creates 200 test Samples and ComputedFiles using the test test Organism and adds them to the test experiment
```
Comment (kurtwheeler) [inline qn_reference.md]:
```suggestion
    - creates 200 test Samples and ComputedFiles using the test Organism and adds them to the test experiment
```

File: qn_reference.md [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    - creates test Dataset with the test experiment and samples 1-6
+    - adds the test Dataset to the ProcessorJob
+- runs QN Reference
+- asserts the job was successfull
```
Comment (kurtwheeler) [inline qn_reference.md]:
```suggestion
- asserts the job was successful
```

File: qn_reference.md [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+Test that the management command fires off and then does not create
+a job for an organism that does not have enough samples on the same
+platform
+
+test steps:
+- creates prerequisites
+    - creates test Organism
+    - creates test experiment
+    - creates 6 test Samples and ComputedFiles using the test test Organism and adds them to the test experiment
```
Comment (kurtwheeler) [inline qn_reference.md]:
```suggestion
    - creates 6 test Samples and ComputedFiles using the test Organism and adds them to the test experiment
```

**Review by kurtwheeler — APPROVED** [review]
The content all looks good to me, so all my comments are spelling/capitalization. Jackie may have additional feedback.

File: no_op.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
+# No Op
```
Comment (jaclyn-taroni) [inline no_op.md:1]: For testing this particular pipeline, I think surfacing any accession codes that we use for "real samples" (if applicable) could be helpful. Submitter-processed data can be all over the place!

File: qn_reference.md [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    - creates test Organism
+    - creates test experiment
+    - creates 6 test Samples and ComputedFiles using the test Organism and adds them to the test experiment
+    - creates test Dataset with the test experiment and samples 1-6
+    - adds the test Dataset to the ProcessorJob
+- calls `create_qn_target` expecting it to fail
+- asserts that the error printed to stdout includes "Target file"
+- asserts that a ProcessorJob wasn't created since there aren't enough samples
+
+## Future Tests
```
Comment (jaclyn-taroni) [inline qn_reference.md:51]: Do we do any testing around NA values? We got burned somewhat badly by including NA values in the QN target calculations in the past.

**Review by jaclyn-taroni — APPROVED** [review]
LGTM. A comment and a question, but nothing that I feel very strongly about re-reviewing!

File: no_op.md (hunk 13 above)
Comment (bug-sam) [reply no_op.md:1]: I'm not sure if I 100% understood this, but I added some additional info to the end of the Future Tests section of No Op. LMK if I am on the right track.

**Pull request merged**